nano/node/online_reps.cpp

```diff
@@ -1,101 +1,131 @@
 	}
 }
 
 void nano::online_reps::observe (nano::account const & rep_a)
 {
 	if (ledger.weight (rep_a) > 0)
 	{
+		auto now = std::chrono::steady_clock::now ();
 		nano::lock_guard<std::mutex> lock (mutex);
-		reps.insert (rep_a);
+		auto new_insert = reps.get<tag_account> ().erase (rep_a) == 0;
+		reps.insert ({ now, rep_a });
+		auto cutoff = reps.get<tag_time> ().lower_bound (now - std::chrono::seconds (config.network_params.node.weight_period));
+		auto trimmed = reps.get<tag_time> ().begin () != cutoff;
+		reps.get<tag_time> ().erase (reps.get<tag_time> ().begin (), cutoff);
+		if (new_insert || trimmed)
+		{
+			online_m = calculate_online ();
+		}
 	}
 }
 
 void nano::online_reps::sample ()
 {
-	auto transaction (ledger.store.tx_begin_write ({ tables::online_weight }));
-	// Discard oldest entries
-	while (ledger.store.online_weight_count (transaction) >= network_params.node.max_weight_samples)
+	nano::unique_lock<std::mutex> lock (mutex);
+	nano::uint128_t online_l;
+	online_l = online_m;
+	lock.unlock ();
+	nano::uint128_t trend_l;
 	{
-		auto oldest (ledger.store.online_weight_begin (transaction));
-		debug_assert (oldest != ledger.store.online_weight_end ());
-		ledger.store.online_weight_del (transaction, oldest->first);
+		auto transaction (ledger.store.tx_begin_write ({ tables::online_weight }));
+		// Discard oldest entries
+		while (ledger.store.online_weight_count (transaction) >= config.network_params.node.max_weight_samples)
+		{
+			auto oldest (ledger.store.online_weight_begin (transaction));
+			debug_assert (oldest != ledger.store.online_weight_end ());
+			ledger.store.online_weight_del (transaction, oldest->first);
+		}
+		ledger.store.online_weight_put (transaction, std::chrono::system_clock::now ().time_since_epoch ().count (), online_l);
+		trend_l = calculate_trend (transaction);
 	}
-	// Calculate current active rep weight
+	lock.lock ();
+	trended_m = trend_l;
+}
+
+nano::uint128_t nano::online_reps::calculate_online () const
+{
 	nano::uint128_t current;
-	std::unordered_set<nano::account> reps_copy;
+	for (auto & i : reps)
 	{
-		nano::lock_guard<std::mutex> lock (mutex);
-		last_reps = reps;
-		reps_copy.swap (reps);
+		current += ledger.weight (i.account);
 	}
-	for (auto & i : reps_copy)
-	{
-		current += ledger.weight (i);
-	}
-	ledger.store.online_weight_put (transaction, std::chrono::system_clock::now ().time_since_epoch ().count (), current);
-	auto trend_l (trend (transaction));
-	nano::lock_guard<std::mutex> lock (mutex);
-	online = trend_l;
+	return current;
 }
 
-nano::uint128_t nano::online_reps::trend (nano::transaction & transaction_a)
+nano::uint128_t nano::online_reps::calculate_trend (nano::transaction & transaction_a) const
 {
 	std::vector<nano::uint128_t> items;
-	items.reserve (network_params.node.max_weight_samples + 1);
-	items.push_back (minimum);
+	items.push_back (config.online_weight_minimum.number ());
+	items.reserve (config.network_params.node.max_weight_samples + 1);
```

Comment:
Most implementations tend to give `std::vector` a default size of 0, so could reduce an allocation swapping these lines back (i.e `reserve` first)